toolkit/tools/imagegen/diskutils/initramfs.go

```diff
@@ -0,0 +1,229 @@
+		if err != nil {
+			return
+		}
+
+		if isLink {
+			var bytesWrittenInt int
+
+			// Write returns an int, cast it to an int64 afterwards
+			logger.Log.Tracef("Creating link %s -> %s", nextFileHeader.Name, nextFileHeader.Linkname)
+			bytesWrittenInt, err = initramfs.cpioWriter.Write(linkPayload)
+			bytesIO = int64(bytesWrittenInt)
+
+		} else {
+			bytesIO, err = io.Copy(initramfs.cpioWriter, cpioReader)
+		}
+
+		if err != nil {
+			return
+		}
+
+		logger.Log.Tracef("File %s caused %d bytes to be transfered to new archive", nextFileHeader.Name, bytesIO)
+		logger.Log.Tracef("Buffer unread length: %d", initramfs.outputBuffer.Len())
+	}
+
+	inputFile.Close()
+
+	fileInfo, err := os.Stat(initramfsPath)
+	if err != nil {
+		return
+	}
+
+	// We can't edit a CPIO archive in place, completely overwrite the file with truncate
+	// The output buffer in memory will be used to re-create the initramfs.
+	initramfs.initramfsOutputFile, err = os.OpenFile(initramfsPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, fileInfo.Mode())
+
+	return
+}
+
+// Close flushes the archives and closes all initramfs resources
+func (i *InitramfsMount) Close() (err error) {
+	var bytesIO int
+
+	logger.Log.Debugf("Closing initramfs file '%s'", i.initramfsOutputFile.Name())
+
+	// Defer close calls to make sure we handle any errors, failing to
+	// close the file means we can't close the install root.
+	defer i.initramfsOutputFile.Close()
+	defer i.pgzWriter.Close()
+	defer i.cpioWriter.Close()
+
+	err = i.cpioWriter.Close()
+	if err != nil {
+		logger.Log.Errorf("Failed to close initramfs: '%s'", err.Error())
+		return
+	}
+	err = i.pgzWriter.Close()
+	if err != nil {
+		logger.Log.Errorf("Failed to close initramfs: '%s'", err.Error())
+		return
+	}
+
+	logger.Log.Debugf("Writing %d bytes to file", i.outputBuffer.Len())
+	bytesIO, err = i.initramfsOutputFile.Write(i.outputBuffer.Bytes())
+	if err != nil {
+		logger.Log.Errorf("Failed to write initramfs file: '%s'", err.Error())
+		return
+	}
+	logger.Log.Infof("Bytes writen to file: %d", bytesIO)
+
+	// Explicit call to fsync, archive corruption was occuring occasionally otherwise.
+	err = i.initramfsOutputFile.Sync()
+	if err != nil {
+		logger.Log.Errorf("Failed to sync initramfs file: '%s'", err.Error())
+		return
+	}
+
+	err = i.initramfsOutputFile.Close()
+	if err != nil {
+		logger.Log.Errorf("Failed to close initramfs: '%s'", err.Error())
+		return
+	}
+	return
+}
+
+// AddFileToInitramfs places a single file in the initramfs at the destination path.
+// - sourcePath: Path to file which is to be adde
+// - destPath: Final destination in the initramfs
+func (i *InitramfsMount) AddFileToInitramfs(sourcePath, destPath string) (err error) {
+	var bytesIO int64
+	fileInfo, err := os.Lstat(sourcePath)
+	file, err := os.Open(sourcePath)
+	if err != nil {
+		return
+	}
+	defer file.Close()
+
+	// Symlinks need to be resolved to their target file to be added to the cpio archive.
+	var link string
```

Comment:
>link [](start = 5, length = 4)

Maybe make the name explaing its purpose a bit more. Is it the link's "source" or "destination" for instance?